### ISSUE TITLE
Detect signed integer overflows

### ISSUE THREAD

**Issue by kees** [KSPP/linux#26]
Much like the array bounds checking (issue #25), the kernel should be detecting signed integer overflows using the Undefined Behavior Sanitizer (UBSan) compiler feature. There are some false positives that need to be fixed (e.g. commit a318f12ed8843cfac53198390c74a565c632f417), and there are some true positives that are expected (e.g. refcount_t overflow), and need to be marked.

There is, however, a complication with the kernel's use of `-fno-strict-overflow` which implies `-fwrapv-pointer` and `-fwrapv`. The former is for wrapped unsigned integer overflow (i.e. unsigned long pointer values), and the latter is needed for true positive signed overflow (i.e. refcount_t). However, this makes integer overflow _no longer undefined behavior_, making UBSan not catch overflows any more. :( To fix this, we need the "intentional overflow/wrap" helpers to DTRT in the face of `-fno-wrapv` so that UBSan will work correctly.

So, things to do:
- create "expected signed overflow" helper inline functions marked with `__attribute__((no_sanitize("signed-integer-overflow")))`
- find all true positives and replace with helper calls
- add note to "deprecated.rst" with something like "open coded signed integer wrap around without a helper"
- split signed integer overflow into a separate Kconfig
- add `-fno-wrapv` back to the signed integer overflow build options

**Comment by kees** [comment]
To illustrate the issue:

```
#include <stdio.h>
#include <limits.h>

/* volatile to avoid optimization */
volatile int val;

int main(void)
{   
    val = INT_MAX; 
    val += 1;
    printf("%d\n", val);
    return 0;
}
```

Normally, the expected warning is generated:
```
$ gcc -Wall -O2 -fsanitize=signed-integer-overflow -o wrap wrap.c
$ ./wrap
wrap.c:10:9: runtime error: signed integer overflow: 2147483647 + 1 cannot be represented in type 'int'
-2147483648
```

But with `-fno-strict-overflow` the condition goes unnoticed:
```
$ gcc -Wall -O2 -fsanitize=signed-integer-overflow -fno-strict-overflow -o wrap wrap.c
$ ./wrap
-2147483648
```

**Comment by kees** [comment]
See also commit 6aaa31aeb9cf260e1b7155cc11ec864f052db5ec

**Comment by kees** [comment]
Interaction with -fno-wrapv: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=102317

**Comment by kees** [comment]
A nice test for the UB (thank you to Miguel Ojeda) here is:
```
bool f(const int x) {
    return x + 1 > x;
}
```

UB will show this as always returning true.

**Comment by kees** [comment]
More notes/background: https://danluu.com/integer-overflow/

**Comment by kees** [comment]
Would have caught:
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=722d94847de2

**Comment by bwendling** [comment]
Copying this GCC documentation here for my benefit:

`-fstrict-overflow`
> Allow the compiler to assume strict signed overflow rules, depending on the language being compiled. For C (and C++) this means that overflow when doing arithmetic with signed numbers is undefined, which means that the compiler may assume that it does not happen. This permits various optimizations. For example, the compiler assumes that an expression like i + 10 > i is always true for signed i. This assumption is only valid if signed overflow is undefined, as the expression is false if i + 10 overflows when using twos complement arithmetic. When this option is in effect any attempt to determine whether an operation on signed numbers overflows must be written carefully to not actually involve overflow.
>
> This option also allows the compiler to assume strict pointer semantics: given a pointer to an object, if adding an offset to that pointer does not produce a pointer to the same object, the addition is undefined. This permits the compiler to conclude that p + u > p is always true for a pointer p and unsigned integer u. This assumption is only valid because pointer wraparound is undefined, as the expression is false if p + u overflows using twos complement arithmetic.
>
> See also the -fwrapv option. Using -fwrapv means that integer signed overflow is fully defined: it wraps. When -fwrapv is used, there is no difference between -fstrict-overflow and -fno-strict-overflow for integers. With -fwrapv certain types of overflow are permitted. For example, if the compiler gets an overflow when doing arithmetic on constants, the overflowed value can still be used with -fwrapv, but not otherwise.
>
> The -fstrict-overflow option is enabled at levels -O2, -O3, -Os.


`-fwrapv`
> This option instructs the compiler to assume that signed arithmetic overflow of addition, subtraction and multiplication wraps around using twos-complement representation. This flag enables some optimizations and disables others. This option is enabled by default for the Java front-end, as required by the Java language specification.

**Comment by kees** [comment]
Also:

`-fwrapv-pointer`

> This option instructs the compiler to assume that pointer arithmetic overflow on addition and subtraction
> wraps around using twos-complement representation.  This flag disables some optimizations which assume pointer
> overflow is invalid.

**Comment by JustinStitt** [comment]
Does anyone know why this sanitizer only works for some integer types and not all? Is this intentional?

See: https://godbolt.org/z/E9MhM7rEb

Here it works fine with `int` but not with `int8_t`.

**Comment by kees** [comment]
Yet another bug! Yes, it seems that it is limited only to 32-bit signed overflow. :(

**Comment by kees** [comment]
Here's a recent thread with GCC folks where we clarified language and I tried to explain the kernel's behavioral needs:
https://gcc.gnu.org/pipermail/gcc-patches/2023-September/630578.html

**Comment by JustinStitt** [comment]
> Yet another bug! Yes, it seems that it is limited only to 32-bit signed overflow. :(

This is likely due to promotion rules; we talked about this offline just wanted to keep this thread up-to-date, somewhat 😸

**Comment by JustinStitt** [comment]
Just dumping some info in this thread.

Here's how I currently see the progress on this issue as of now:

(note that these commit SHAs mentioned are from Kees' tree linked below and may change under force push / rebase circumstances).

> * [x]  make signed integer sanitizer work even with `-fwrapv` (and `-fno-strict-overflow`).
https://github.com/llvm/llvm-project/pull/82432

> * [ ]  create "expected signed overflow" helper inline functions marked with `__attribute__((no_sanitize("signed-integer-overflow")))`..

https://git.kernel.org/pub/scm/linux/kernel/git/kees/linux.git/log/?h=dev/v6.8-rc2/signed-overflow-sanitizer

with Commit e30a6c0ba369056b ("overflow: Introduce wrapping_add(), wrapping_sub(), and wrapping_mul()") and Commit 87e7e797d6a0fbb8 ("overflow: Introduce wrapping_assign_add() and wrapping_assign_sub()")

> * [ ]  find all true positives and replace with helper calls.

"all" is a big word. hard to mark this as done.

https://git.kernel.org/pub/scm/linux/kernel/git/kees/linux.git/log/?h=dev/v6.8-rc2/signed-overflow-sanitizer

main offenders: refcount and atomics covered in tree with commits:

* `4d6a54c480d91c0c9210d1041342a38ac840d4a7`

* `b037d9207c40cdae50424de2d9281bcfde9f3485`

* `53003a74bdc005a4517de0ab0edca70ba28b0ed`

* `1ebd18c24a127e5d7c56f9d888e2b3ad8c8e9f32`

> * [ ]  add note to "deprecated.rst" with something like "open coded signed integer wrap around without a helper".
^ todo

> * [x]  add back signed integer overflow as a UBSan Kconfig


https://git.kernel.org/pub/scm/linux/kernel/git/kees/linux.git/log/?h=dev/v6.8-rc2/signed-overflow-sanitizer

Done with 557f8c582a9ba8abe6aa0fd734b6f342af106b26

**Comment by kees** [comment]
Yes! I've updated the top-level now (and added you to the project ACLs to edit this for yourself in the future).
